[Excel](data types) How to create lambda methods (#5128)
OfficeDev/office-js-docs-pr: docs/excel/excel-add-ins-dot-functions.md

---
title: Add lambda methods to cell values
description: Add lambda methods to cell values to support dot syntax in Excel formulas.
ms.date: 04/16/2022
ms.localizationpriority: medium
---

# Add lambda methods to cell values

> Add lambda methods to cell values to provide the user access to dynamic calculations based on the cell value. You can add lambda methods to the `EntityCellValue` and `LinkedEntityCellValue` types. For example, you can add a method to a product entity value that converts its weight to different units.

Reviewer: ```suggestion
Add lambda methods to cell values to provide the user access to dynamic calculations based on the cell value. You can add lambda methods to the `EntityCellValue` and `LinkedEntityCellValue` types. For example, add a method to a product entity value that converts its weight to different units.
```
There are 20 "you cans" in this article, going to try to trim a few!

The following screenshot shows an example of adding a `ConvertWeight` method to a product entity value representing pancake mix.

:::image type="content" source="/images/excel-add-in-dot-function.png" alt-text="Screenshot of Excel formula showing =A1.ConvertWeight( ounces )":::

You can also add lambda methods to the `DoubleCellValue`, `BooleanCellValue`, and `StringCellValue` types. The following screenshot shows an example of adding a `ConvertToRomanNumeral` method to a double value type.

:::image type="content" source="/images/excel-add-in-dot-function-roman-numeral.png" alt-text="Screenshot of Excel formula showing =A1.ConvertToRomanNumeral()":::

Lambda methods don’t appear on the data type card for the user.

:::image type="content" source="/images/excel-add-in-dot-function-data-card.png" alt-text="Screenshot of data card for Pancake mix data type, but no lamdba methods are listed.":::

## Add a lambda method to an entity value

> To add a lambda method to an entity value, describe it in JSON using the `LambdaCellValue` type. The following example shows how to define a simple method that returns the value 27.

Reviewer: ```suggestion
To add a lambda method to an entity value, describe it in JSON using the `LambdaCellValue` type. The following code sample shows how to define a simple method that returns the value 27.
```

The following code sample shows the implementation of the `Contoso.AddValue` function.

```typescript
/**
 * Adds x to the calling object.
 * @customfunction
 * @excludeFromAutocomplete
 * @param {any} math The math object (calling object).
 * @param {number} x The value to add.
 * @return {number[][]}  Sum.
 */
function addValue(math, x): number[][] {  
  const result: number = math.properties["value"].basicValue + x;
  return [[result]];
}
```

Note the following about the previous code sample.

- The `@excludeFromAutocomplete` tag ensures the custom method doesn't appear to the user in the Excel UI when entering it in a search box. However, note that a user can still call the custom function separately from an entity value if they enter it directly into a cell.
- The calling object is always passed as the first argument and must by of type `any`. In this case, it's named `math` and is used to get the value property from the `math` object.
- It returns a double array of numbers.

> When the user interacts with the reference method in Excel, they don’t see the calling object as an argument.

Author: yes

  const entity: Excel.EntityCellValue = {
    type: Excel.CellValueType.entity,
    text: productName,
    properties: {
      "Product ID": {
        type: Excel.CellValueType.string,
        basicValue: productID.toString() || ""
      },
      "Product Name": {
        type: Excel.CellValueType.string,
        basicValue: productName || ""
      },
      "Unit Price": {
        type: Excel.CellValueType.formattedNumber,
        basicValue: price,
        numberFormat: "$* #,##0.00"
      },
      applySalesTax: referenceCustomFunctionCalculateSalesTax
    },
  };
  return entity;
}
```

## Exclude custom functions from the Excel UI

We recommend you use the `@excludeFromAutoComplete` tag in the comments description of custom functions used by reference methods. It indicates that the function will be excluded from the autocomplete drop-down list and Formula Builder. This helps prevent the user from accidentally using a custom function separately from its entity value.

> [!NOTE]
> If the function is manually entered correctly in the grid, the function will still execute. Also, a function can’t have both `@excludeFromAutoComplete` and `@linkedEntityLoadService` tags.

> For the full list of properties in addition to the properties specified by [Manually create JSON metadata for custom functions](custom-functions-json.md) the `excludeFromAutoComplete` property is available for reference methods.

Author: That sentence should be deleted.

The following example shows how to manually describe the `APPLYSALESTAX` with JSON in the **functions.json** file. The `excludeFromAutoComplete` property is set to `true`.

```typescript
{
    "description": "Calculates the price when a sales tax rate is applied.",
    "id": "APPLYSALESTAX",
    "name": "APPLYSALESTAX",
    "options": {
        "excludeFromAutoComplete": true,
        "capturesCallingObject": true
    },
    "parameters": [
        {
            "description": "The product entity value (calling object).",
            "name": "product",
            "type": "any"
        },
        {
            "description": "The tax rate (0.11 = 11%).",
            "name": "taxRate",
            "type": "number"
        }
    ],
    "result": {
        "dimensionality": "matrix",
        "type": "number"
    }
},
```

For more information, see [Manually create JSON metadata for custom functions](custom-functions-json.md).

## Add a function to a basic value type

> You can add functions to the basic value types of `Boolean`, `double`, and `string`. The process is the same as for entity values. Describe the function with JSON as a reference method. The following code sample shows how to create a double basic value with a function `AddValue()` that adds a value `x` to the basic value.

Reviewer: ```suggestion
To add functions to the basic value types of `Boolean`, `double`, and `string`, the process is the same as for entity values. Describe the function with JSON as a reference method. The following code sample shows how to create a double basic value with a function `AddValue()` that adds a value `x` to the basic value.
```

}
```

The following code sample shows the implementation of the reference method as a custom function named `generateRandomRange`. It returns a dynamic array of random values matching the number of `rows` and `columns` specified. The `min` and `max` arguments are optional, and if not specified will default to `1` and `10`.

```typescript
/**
 * Generate a dynamic array of random numbers.
 * @customfunction
 * @excludeFromAutocomplete
 * @param {number} rows Number of rows to generate.
 * @param {number} columns Number of columns to generate.
 * @param {number} [min] Lowest number that can be generated. Default is 1.
 * @param {number} [max] Highest number that can be generated. Default is 10.
 * @returns {number[][]} A dynamic array of random numbers.
 */
function generateRandomRange(rows, columns, min, max) {
  // Set defaults for any missing optional arguments.
  if (min===null) min = 1;
  if (max === null) max = 10;

  let numbers = new Array(rows);
  for (let r = 0; r < rows; r++) {
    numbers[r] = new Array(columns);
    for (let c = 0; c < columns; c++) {
      numbers[r][c] = Math.round(Math.random() * (max - min) ) + min;
    }
  }
  return numbers;
}
```

> As the user enters the function in Excel, autocomplete shows the properties of the function, and indicates optional arguments by surrounding them in brackets []. The following image shows an example of entering optional parameters using the `generateRandomRange` reference method.

Reviewer: ```suggestion
When the user enters the function in Excel, autocomplete shows the properties of the function, and indicates optional arguments by surrounding them in brackets []. The following image shows an example of entering optional parameters using the `generateRandomRange` reference method.
```

:::image type="content" source="../images/excel-data-types-reference-method-optional-parameters.png" alt-text="Screenshot of entering concatProductNames method in Excel passing A1 and A2 which contain a bicycle and unicycle product entity value.":::

> ## Repeating parameters

Reviewer: ```suggestion
## Multiple parameters
```
I'm not sure if this is clear. "repeating parameters" to me means the same parameter, repeated. It does not mean an array of different product names. Do you mean "multiple parameters?"

Author: They are indeed the same parameter as they are all "products".

> Add repeating parameters to a function by using an array. The following code sample shows how to create a reference function that concatenates zero or more product names passed in a products array.

Reviewer: ```suggestion
Add multiple parameters to a function by using an array. The following code sample shows how to create a reference function that concatenates zero or more product names passed in a products array.
```

```typescript
/** 
 * @customfunction 
 * @excludeFromAutocomplete 
 * @description Concatenate the names of given products, joined by " | " 
 * @param {any[]} products - The products to concatenate. 
 * @returns A string of concatenated product names. 
 */ 
function concatProductNames(products: any[]): string { 
  return products.map((product) => product.properties["Product Name"].basicValue).join(" | "); 
} 

```

The following code sample shows how to create an entity with the `concatProductNames` reference method.

```typescript
const referenceCustomFunctionRepeating: Excel.JavaScriptCustomFunctionReferenceCellValue = { 
  type: Excel.CellValueType.function,
  functionType: Excel.FunctionCellValueType.javaScriptReference,
  namespace: "CONTOSO", 
  id: "CONCATPRODUCTNAMES" 
} 

function makeProductEntity(productID: number, productName: string, price: number) {
  const entity: Excel.EntityCellValue = {
    type: Excel.CellValueType.entity,
    text: productName,
    properties: {
      "Product ID": {...},
      "Product Name": {...},
      "Unit Price": {...},
      concatProductNames: referenceCustomFunctionRepeating,
    },
  };
  return entity;
}
```

> The following image shows an example of entering repeating parameters using the `concatProductNames` reference method.

Reviewer: ```suggestion
The following image shows an example of entering multiple parameters using the `concatProductNames` reference method.
```

```typescript
/** 
 * @customfunction 
 * @excludeFromAutocomplete 
 * @description Calculate the sum of arbitrary parameters. 
 * @param {number[][][]} operands - The operands to sum. 
 * @returns  The sum of all operands. 
 */ 
function sumAll(operands: number[][][]): number { 
  let total: number = 0; 
 
  operands.forEach(range => { 
    range.forEach(row => { 
      row.forEach(num => { 
        total += num; 
      }); 
    }); 
  }); 
 
  return total; 
} 
```

The following code sample shows how to create an entity with the `sumAll` reference method.

```typescript
const referenceCustomFunctionRange: Excel.JavaScriptCustomFunctionReferenceCellValue = { 
  type: Excel.CellValueType.function,
  functionType: Excel.FunctionCellValueType.javaScriptReference,
  namespace: "CONTOSO", 
  id: "SUMALL" 
} 

function makeProductEntity(productID: number, productName: string, price: number) {
  const entity: Excel.EntityCellValue = {
    type: Excel.CellValueType.entity,
    text: productName,
    properties: {
      "Product ID": {...},
      "Product Name": {...},
      "Unit Price": {...},
      sumAll: referenceCustomFunctionRange
    },
  };
  return entity;
}
```

> The following image shows an example of entering repeating parameters with a range parameter using the `sumAll` reference method.

Reviewer: ```suggestion
The following image shows an example of entering multiple parameters with a range parameter using the `sumAll` reference method.
```
